apache/rocketmq/v1/service.proto

```diff
@@ -0,0 +1,382 @@
+  }
+  Source source = 5;
+  string trace_context = 6;
+
+  reserved 7 to 64;
+}
+
+message EndTransactionResponse {
+  ResponseCommon common = 1;
+
+  reserved 2 to 64;
+}
+
+message QueryOffsetRequest {
+  Partition partition = 1;
+  QueryOffsetPolicy policy = 2;
+  google.protobuf.Timestamp time_point = 3;
+
+  reserved 4 to 64;
+}
+
+message QueryOffsetResponse {
+  ResponseCommon common = 1;
+  int64 offset = 2;
+
+  reserved 3 to 64;
+}
+
+message PullMessageRequest {
+  Resource group = 1;
+  Partition partition = 2;
+  int64 offset = 3;
+  int32 batch_size = 4;
+  google.protobuf.Duration await_time = 5;
+  FilterExpression filter_expression = 6;
+  string client_id = 7;
+
+  reserved 8 to 64;
+}
+
+message PullMessageResponse {
+  ResponseCommon common = 1;
+  int64 min_offset = 2;
+  int64 next_offset = 3;
+  int64 max_offset = 4;
+  repeated Message messages = 5;
+
+  reserved 6 to 64;
+}
+
+message UpdateOffsetRequest {
+  Resource group = 1;
+  Partition partition = 2;
+  int64 partition_offset = 3;
+
+  reserved 4 to 64;
+}
+
+message UpdateOffsetResponse {
+  ResponseCommon common = 1;
+
+  reserved 2 to 64;
+}
+
+message GenericPollingRequest {
+  string client_id = 1;
+  repeated Resource topics = 2;
+  oneof group {
+    Resource producer_group = 3;
+    Resource consumer_group = 4;
+  }
+
+  reserved 5 to 64;
+}
+
+message GenericPollingResponse {
+  ResponseCommon common = 1;
+
+  reserved 2 to 64;
+}
+
+message PrintThreadStackRequest {
+  string mid = 1;
+
+  reserved 2 to 64;
+}
+
+message PrintThreadStackResponse {
+  ResponseCommon common = 1;
+  string mid = 2;
+  string stack_trace = 3;
+
+  reserved 4 to 64;
+}
+
+message VerifyMessageConsumptionRequest {
+  string mid = 1;
+  Message message = 2;
+
+  reserved 3 to 64;
+}
+
+message VerifyMessageConsumptionResponse {
+  string mid = 1;
+  ResponseCommon common = 2;
+
+  reserved 3 to 64;
+}
+
+message RecoverOrphanedTransactionRequest {
+  Message orphaned_transactional_message = 1;
+  string transaction_id = 2;
+
+  reserved 3 to 64;
+}
+
+message MultiplexingRequest {
+  oneof type {
+    GenericPollingRequest polling_request = 1;
+    PrintThreadStackResponse print_thread_stack_response = 2;
+    VerifyMessageConsumptionResponse verify_message_consumption_response = 3;
+  }
+
+  reserved 4 to 64;
+}
+
+message MultiplexingResponse {
+  oneof type {
+    GenericPollingResponse polling_response = 1;
+    PrintThreadStackRequest print_thread_stack_request = 2;
+    VerifyMessageConsumptionRequest verify_message_consumption_request = 3;
+    RecoverOrphanedTransactionRequest recover_orphaned_transaction_request = 4;
+  }
+
+  reserved 5 to 64;
+}
+
+message NotifyClientTerminationRequest {
+  Resource group = 1;
+  string client_id = 2;
+
+  reserved 3 to 64;
+}
+
+message NotifyClientTerminationResponse {
+  ResponseCommon common = 1;
+
+  reserved 2 to 64;
+}
+
+service MessagingService {
+  rpc QueryRoute(QueryRouteRequest) returns (QueryRouteResponse) {}
```

Comment:
Service items should be properly documented, describing their usage, context, and more importantly, error handling.

A typical example for reference: https://github.com/envoyproxy/data-plane-api/blob/main/envoy/api/v2/discovery.proto

Reply:
in progress~